### ISSUE TITLE
Question: After user_ta_enter_open_session(), call tee_ta_clear_busy() with TA_FLAG_SINGLE_INSTANCE

### ISSUE THREAD

**Issue by HeJyunYan** [OP-TEE/optee_os#4471]
In commit [f0ab1c64](https://github.com/OP-TEE/optee_os/commit/f0ab1c64777339e3ae508eb13347e0e4518b5d03), the `init_with_ldelf` call is moved from `user_ta_enter_open_session` to `tee_ta_init_session`.
It means that the ctx is fully initialized after `tee_ta_init_session` is done.
Thus when open a new session to a single instance TA first time, `lock_single_instance` will be called since the ctx->flags is already set.
But the `unlock_single_instance` won't be called in `tee_ta_clear_busy` (ctx->initializing is still false).
It results that the `tee_ta_single_instance_count` will always increase once open a session to a single instance TA first time.
This makes the `tee_ta_single_instance_thread` won't be reset to `THREAD_ID_INVALID` since the `tee_ta_single_instance_count` leaks problem described above.
Therefore, the latter thread may be blocked in `tee_ta_try_set_busy` once the thread ID is not the same as the previous thread.
Does the variable "`initializing`" is unused and can be removed from ctx?
And should the condition "`if (!ctx->initializing && (ctx->flags & TA_FLAG_SINGLE_INSTANCE))`" in `tee_ta_clear_busy` be changed to "`if (ctx->flags & TA_FLAG_SINGLE_INSTANCE)`"?

**Comment by HeJyunYan** [comment]
The issue we found is that when creating a new ctx for a single instance TA, lock_single_instance will be called in tee_ta_init_user_ta_session().
However unlock_single_instance is not since the ctx->initializing is "true" before it's be cleared in tee_ta_clear_busy().

The variable "ctx->initializing" is added in commit [d1911a85](https://github.com/OP-TEE/optee_os/commit/d1911a85142da16fef5ebdcdac0348d29ce37cd8).
This variable looks like a patch added in tee_ta_try_set_busy / tee_ta_clear_busy since ctx->flags has not been initialized.
Thus lock_single_instance won't be called even a single instance TA at this moment since the ctx->flags cannot be relied in tee_ta_try_set_busy.
(init_with_ldelf is not called yet).
This commit uses ctx->initializing to avoid calling unlock_single_instance() in tee_ta_clear_busy in this situation, .

But after commit [f0ab1c64](https://github.com/OP-TEE/optee_os/commit/f0ab1c64777339e3ae508eb13347e0e4518b5d03), ctx->flags can already be relied after tee_ta_init_session is done (since init_with_ldelf is moved here from ctx->ops->enter_open_session, ctx->flags is fully initialized).
Therefore lock_single_instance() will be called if it's an single instance TA.

Thus we think the condition "ctx->initializing" should also be removed from tee_ta_clear_busy.

Is this right?

**Comment by jenswi-linaro** [comment]
I'm sorry I can't quite see where it goes wrong. Does this relate to https://github.com/OP-TEE/optee_os/pull/4472 ?
If there's an issue here or not I think we need to focus on what's in the master branch right now.

**Comment by HeJyunYan** [comment]
No It's different issues.
Below is the part of code section when one CA opens a session to a single instance TA,
```
TEE_Result tee_ta_open_session( ... )
			       uint32_t cancel_req_to,
			       struct tee_ta_param *param)
{
    ...

	res = tee_ta_init_session(err, open_sessions, uuid, &s);
	/*
	    code flow:
	    tee_ta_init_session -> 
	    tee_ta_init_user_ta_session ->
	    ldelf_init_with_ldelf ->
	    ldelf_init_with_ldelf
	    ctx->flags is set in ldelf_init_with_ldelf
	    and ctx->initializing is true here
	*/
	if (res != TEE_SUCCESS) {
		DMSG("init session failed 0x%x", res);
		return res;
	}

    ...

	if (tee_ta_try_set_busy(ctx)) {
		s->param = param;
		set_invoke_timeout(s, cancel_req_to);
		res = ts_ctx->ops->enter_open_session(&s->ts_sess);
		tee_ta_clear_busy(ctx);
	} else {
		/* Deadlock avoided */
		res = TEE_ERROR_BUSY;
		was_busy = true;
	}

    ...

	return res;
}

```

```
static bool tee_ta_try_set_busy(struct tee_ta_ctx *ctx)
{
	...
	/*
	    ctx->flags is not 0, this condition is true when single instance TA
	    so lock_single_instance is called here
	*/
	if (ctx->flags & TA_FLAG_SINGLE_INSTANCE)
		lock_single_instance();
	...
	return rc;
}
```

```
static void tee_ta_clear_busy(struct tee_ta_ctx *ctx)
{
	...
	/*
	    ctx->initializing is still false, but lock_single_instance has been called before
	    No call to unlock_single_instance, it results that tee_ta_single_instance_count leak
	*/
	if (!ctx->initializing && (ctx->flags & TA_FLAG_SINGLE_INSTANCE))
		unlock_single_instance();

	ctx->initializing = false;
	...
}
```

**Comment by jenswi-linaro** [comment]
OK, I agree there seem to be some problem here. Do you think you can propose a fix for this? Perhaps this is better done after https://github.com/OP-TEE/optee_os/pull/4472 because it seems there's some overlap at least, in the state variables if nothing else.

**Comment by HeJyunYan** [comment]
Fixed by commit [c571b3f](https://github.com/OP-TEE/optee_os/commit/c571b3fd4956d70c8f4435f94d64c7e36c25885d)
